File: qa/p5-entity-ssg-seo-technical-check.py
#!/usr/bin/env python3
"""P5-E.9F.2 — Python fallback for p5-entity-ssg-seo-technical-check.mjs"""
import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
FIXTURE = ROOT / "qa" / "fixtures" / "entity-ssg-fixtures.json"
SITEMAP = ROOT / "qa" / "entity-ssg-sitemap.fixture.xml"
PROD_SITEMAP = ROOT / "sitemap.xml"
NOT_FOUND = ROOT / "wiki" / "post" / "_ssg-not-found" / "index.html"
FORBIDDEN = re.compile(
    r"BLMETA|search_text|search_vector|service_role|SUPABASE_SERVICE_ROLE|password|passwd|secret|token|"
    r"@[a-z0-9.-]+\.[a-z]{2,}|\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\b",
    re.I,
)
MARKER = re.compile(r"\b(qa-ssg|p5-e9e|fixture-only|prototype fixture|test fixture)\b", re.I)
SLUG_BLOCK = re.compile(r"^(qa|test|p5|fixture|draft|pending|deleted|contribution)")
pass_count = 0


def fail(msg: str) -> None:
    print(f"[p5-entity-ssg-seo-technical-check] FAIL: {msg}", file=sys.stderr)
    sys.exit(1)


def ok(msg: str) -> None:
    global pass_count
    pass_count += 1
    print(f"[p5-entity-ssg-seo-technical-check] PASS: {msg}")


def check_entity(entity: dict) -> None:
    slug = entity["canonical_slug"]
    html_path = ROOT / "wiki" / "post" / slug / "index.html"
    if not html_path.exists():
        fail(f"Missing output for {slug}")
    html = html_path.read_text(encoding="utf-8")
    checks = [
        ("generator marker", "build-entity-ssg-fixtures.mjs" in html),
        ("title", "<title>" in html and entity["title"] in html),
        ("description", 'meta name="description"' in html),
        ("canonical", 'rel="canonical"' in html),
        ("og:title", 'property="og:title"' in html),
        ("og:description", 'property="og:description"' in html),
        ("og:url", 'property="og:url"' in html),
        ("twitter:title", 'name="twitter:title"' in html),
        ("twitter:description", 'name="twitter:description"' in html),
        ("single h1", len(re.findall(r"<h1\b", html, re.I)) == 1),
        ("excerpt", entity["excerpt"][:40] in html),
        ("no loading", "Loading post" not in html),
        ("no leak", not FORBIDDEN.search(html)),
        ("no markers", not MARKER.search(html)),
        ("CreativeWork", "CreativeWork" in html),
        ("BreadcrumbList", "BreadcrumbList" in html),
        ("noindex", 'meta name="robots" content="noindex, follow"' in html),
    ]
    for name, passed in checks:
        if not passed:
            fail(f"{slug} failed check: {name}")
    ok(f"Validated entity page {slug}")


def main() -> None:
    if not (ROOT / "scripts" / "build-entity-ssg-fixtures.mjs").exists():
        fail("Generator script missing")
    ok("Generator script exists")
    entities = json.loads(FIXTURE.read_text(encoding="utf-8")).get("entities", [])
    if len(entities) < 6:
        fail(f"Expected at least 6 entities, got {len(entities)}")
    ok(f"Fixture count >= 6 ({len(entities)})")
    slugs = set()
    for entity in entities:
        slug = entity["canonical_slug"]
        if slug in slugs:
            fail(f"Duplicate slug {slug}")
        slugs.add(slug)
        if SLUG_BLOCK.match(slug) or not re.match(r"^[a-z0-9]+(?:-[a-z0-9]+)*$", slug):
            fail(f"Invalid slug {slug}")
    ok("All slugs unique and SEO-safe")
    sitemap = SITEMAP.read_text(encoding="utf-8")
    for entity in entities:
        if entity["canonical_url"] not in sitemap:
            fail(f"Sitemap missing {entity['canonical_url']}")
    ok("Fixture sitemap includes all entity URLs")
    prod = PROD_SITEMAP.read_text(encoding="utf-8")
    for entity in entities:
        if f"/wiki/post/{entity['canonical_slug']}/" in prod:
            fail(f"Production sitemap must not include {entity['canonical_slug']}")
    ok("Production sitemap.xml unchanged for fixture entities")
    for entity in entities:
        check_entity(entity)
    nf = NOT_FOUND.read_text(encoding="utf-8")
    if 'meta name="robots" content="noindex, nofollow"' not in nf:
        fail("Not-found page must be noindex,nofollow")
    ok("Not-found fail-closed page present")
    print(f"[p5-entity-ssg-seo-technical-check] All checks passed ({pass_count} assertions)")


if __name__ == "__main__":
    main()

File: qa/p5-entity-ssg-seo-technical-check.mjs
#!/usr/bin/env node
/**
 * P5-E.9F.2 — Node-side check for entity SSG SEO technical implementation.
 * No browser. No network. No DB.
 */

import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, "..");
const FIXTURE_PATH = path.join(ROOT, "qa", "fixtures", "entity-ssg-fixtures.json");
const GENERATOR_PATH = path.join(ROOT, "scripts", "build-entity-ssg-fixtures.mjs");
const SITEMAP_PATH = path.join(ROOT, "qa", "entity-ssg-sitemap.fixture.xml");
const PRODUCTION_SITEMAP = path.join(ROOT, "sitemap.xml");
const NOT_FOUND_PATH = path.join(ROOT, "wiki", "post", "_ssg-not-found", "index.html");

const FORBIDDEN_RE =
  /BLMETA|search_text|search_vector|service_role|SUPABASE_SERVICE_ROLE|password|passwd|secret|token|@[a-z0-9.-]+\.[a-z]{2,}|\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\b/i;
const MARKER_RE = /\b(qa-ssg|p5-e9e|fixture-only|prototype fixture|test fixture)\b/i;
const SLUG_BLOCKLIST = /^(qa|test|p5|fixture|draft|pending|deleted|contribution)/;

let passCount = 0;

function fail(message) {
  console.error("[p5-entity-ssg-seo-technical-check] FAIL:", message);
  process.exit(1);
}

function ok(message) {
  passCount += 1;
  console.log("[p5-entity-ssg-seo-technical-check] PASS:", message);
}

function read(filePath) {
  return fs.readFileSync(filePath, "utf8");
}

function assertHtmlEntity(entity) {
  const slug = entity.canonical_slug;
  const htmlPath = path.join(ROOT, "wiki", "post", slug, "index.html");
  if (!fs.existsSync(htmlPath)) fail("Missing output for " + slug);
  const html = read(htmlPath);
  const lower = html.toLowerCase();

  if (!html.includes("build-entity-ssg-fixtures.mjs")) fail(slug + " missing generator marker");
  if (!html.includes("<title>")) fail(slug + " missing title");
  if (!html.includes('meta name="description"')) fail(slug + " missing description meta");
  if (!html.includes('rel="canonical"')) fail(slug + " missing canonical");
  if (!html.includes('property="og:title"')) fail(slug + " missing og:title");
  if (!html.includes('property="og:description"')) fail(slug + " missing og:description");
  if (!html.includes('property="og:url"')) fail(slug + " missing og:url");
  if (!html.includes('name="twitter:title"')) fail(slug + " missing twitter:title");
  if (!html.includes('name="twitter:description"')) fail(slug + " missing twitter:description");
  if ((html.match(/<h1\b/gi) || []).length !== 1) fail(slug + " must have exactly one H1");
  if (!html.includes(entity.title)) fail(slug + " missing entity title in HTML");
  if (!html.includes(entity.excerpt.slice(0, 40))) fail(slug + " missing excerpt in HTML");
  if (html.includes("Loading post")) fail(slug + " must not contain Loading post");
  if (FORBIDDEN_RE.test(html)) fail(slug + " forbidden leak pattern");
  if (MARKER_RE.test(html)) fail(slug + " QA/test/fixture marker in output HTML");
  if (!html.includes("CreativeWork")) fail(slug + " missing CreativeWork JSON-LD");
  if (!html.includes("BreadcrumbList")) fail(slug + " missing BreadcrumbList JSON-LD");
  if (!html.includes('meta name="robots" content="noindex, follow"')) {
    fail(slug + " must use prelaunch noindex,follow");
  }
  ok("Validated entity page " + slug);
}

function main() {
  if (!fs.existsSync(GENERATOR_PATH)) fail("Generator script missing");
  ok("Generator script exists");

  const fixture = JSON.parse(read(FIXTURE_PATH));
  const entities = fixture.entities || [];
  if (entities.length < 6) fail("Expected at least 6 fixture entities, got " + entities.length);
  ok("Fixture count >= 6 (" + entities.length + ")");

  const slugs = new Set();
  entities.forEach(function (entity) {
    if (slugs.has(entity.canonical_slug)) fail("Duplicate slug " + entity.canonical_slug);
    slugs.add(entity.canonical_slug);
    if (SLUG_BLOCKLIST.test(entity.canonical_slug)) {
      fail("Blocked slug " + entity.canonical_slug);
    }
    if (!/^[a-z0-9]+(?:-[a-z0-9]+)*$/.test(entity.canonical_slug)) {
      fail("Invalid slug " + entity.canonical_slug);
    }
  });
  ok("All slugs unique and SEO-safe");

  if (!fs.existsSync(SITEMAP_PATH)) fail("Fixture sitemap missing");
  const sitemap = read(SITEMAP_PATH);
  entities.forEach(function (entity) {
    if (!sitemap.includes(entity.canonical_url)) {
      fail("Fixture sitemap missing " + entity.canonical_url);
    }
  });
  ok("Fixture sitemap includes all entity URLs");

  const excluded = ["qa/", "test", "draft", "pending", "deleted", "contribution"];
  excluded.forEach(function (token) {
    if (sitemap.toLowerCase().includes("/wiki/post/" + token)) {
      fail("Fixture sitemap contains excluded token: " + token);
    }
  });
  ok("Fixture sitemap excludes private/QA/test patterns");

  const prodSitemap = read(PRODUCTION_SITEMAP);
  entities.forEach(function (entity) {
    if (prodSitemap.includes("/wiki/post/" + entity.canonical_slug + "/")) {
      fail("Production sitemap.xml must not include fixture entity " + entity.canonical_slug);
    }
  });
  ok("Production sitemap.xml unchanged for fixture entities");

  entities.forEach(assertHtmlEntity);

  if (!fs.existsSync(NOT_FOUND_PATH)) fail("Not-found page missing");
  const notFound = read(NOT_FOUND_PATH);
  if (!notFound.includes('meta name="robots" content="noindex, nofollow"')) {
    fail("Not-found page must be noindex,nofollow");
  }
  ok("Not-found fail-closed page present");

  console.log("[p5-entity-ssg-seo-technical-check] All checks passed (" + passCount + " assertions)");
}

main();
